Allow special tags to be self-closing

ang-zeyu opened the pull request:
**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [x] Documentation update
• [x] Enhancement to an existing feature

Resolves #1099

**What is the rationale for this request?**
Allow injected special tags to be self-closing as well, while preserving the contrary for `script/style` tags ( according to html5 spec ).

**What changes did you make? (Give an overview)**


- Simple 2 line patch to self-closing tag state handler to allow said behaviour
```js
// self-closing tag state handler
...
this._special = SPECIAL_NONE;
...
```
- Update tests to test for self-closing injected special tags as well
- Small documentation update to reflect this information

**Is there anything you'd like reviewers to focus on?**
na

**Testing instructions:**
- `npm run test`

**Proposed commit message: (wrap lines at 72 characters)**
Allow special tags to be self-closing

Unlike the html script and style special tags, plugin authors may want
injected special tags to be able to be self-closing as well.
This can help unify the syntax used in the plugin, which makes for
better author usability of the plugin.

Let’s allow injected special tags to be self-closing.
Let’s also do so for the script and style tags, which would be
expanded to their normal form after rendering.

crphang commented on src/lib/markbind/src/patches/htmlparser2.js:
```diff
@@ -261,6 +261,28 @@ Tokenizer.prototype._stateBeforeSpecial = function(c) {
+		this._sectionStart = this._index + 1;
+		if (this._special > 2) {
+			/*
+			 We intentionally fail if the special tag in question is
```
Minor: Would really help to document/comment what "fail" imply when you set `this._special = SPECIAL_NONE`.

ang-zeyu replied:
Updated

ang-zeyu commented:
> Since there's no such thing as self closing tags in the HTML specs, should we just allow script and style tags to be self closing too?
> 
> In the generated HTML they would all be expanded out to the valid `<script></script>` and `<special></special>` forms anyway.

I'm open to either; What do you think? @crphang

crphang commented:
> > Since there's no such thing as self closing tags in the HTML specs, should we just allow script and style tags to be self closing too?
> > In the generated HTML they would all be expanded out to the valid `<script></script>` and `<special></special>` forms anyway.
> 
> I'm open to either; What do you think? @crphang

Looks fine to me, not 100% sure if there will be no issues, but seems like it will make the code simpler here

ang-zeyu commented:
Updated to allow `script / style` tags as well;
It should be side effect free as htmlparser2 dosen't do special processing for script/style tags outside of its tokenizer

crphang commented:
@marvinchin could you help with the review of this? 🙇

marvinchin commented on src/lib/markbind/src/patches/htmlparser2.js:
```diff
@@ -261,6 +261,22 @@ Tokenizer.prototype._stateBeforeSpecial = function(c) {
 };
 
+/**
+ * Removes the special state if the special tag was self-closed.
```
Can we explicitly state that this is an altered version of the `_stateInSelfClosingTag` defined in `htmlparser`? I think it might help other developers new to this part of the codebase understand what's going on better.

marvinchin commented on src/lib/markbind/src/patches/htmlparser2.js:
```diff
@@ -261,6 +261,22 @@ Tokenizer.prototype._stateBeforeSpecial = function(c) {
+		this._cbs.onselfclosingtag();
+		this._state = TEXT;
+		this._sectionStart = this._index + 1;
+		// Allow all special tags (including script & style tags) to be self-closed
```
Let's explain why we allow this despite the spec not allowing it. I think just stating the fact that the original htmlparser allows it is sufficient.

ang-zeyu replied:
To clarify, did you mean the `original htmlparser disallows it` ( that is the case, we're breaking away from the parser's default behaviour here )?

marvinchin replied:
Oops, I actually thought that the original htmlparser allowed it, since the changes we made didn't seem to affect any of the logic specific to script/style tags. After taking a closer look at the rest of the changes in our patches, it seems like you are right 🙂

ang-zeyu commented:
> (even for re-reviews!)

Will do 👍 Thanks for looking through this!

ang-zeyu commented:
Resolved minor conflict with #1049

ang-zeyu commented:
> broken currently

Fixed, updated #1049's cheerio test site call with `xmlMode: false`.
Needed because in xmlMode `<abc></abc>` is recognised as an empty tag, hence it renders back to `<abc />`, which breaks once passed to the next plugin where `xmlMode = false`
